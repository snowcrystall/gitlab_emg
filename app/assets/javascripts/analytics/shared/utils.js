export const filterBySearchTerm = (data = [], searchTerm = '', filterByKey = 'name') => {
  if (!searchTerm?.length) return data;
  return data.filter((item) => item[filterByKey].toLowerCase().includes(searchTerm.toLowerCase()));
};
