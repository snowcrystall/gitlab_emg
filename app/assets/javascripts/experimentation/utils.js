// This file only applies to use of experiments through https://gitlab.com/gitlab-org/gitlab-experiment
import { get } from 'lodash';
import { DEFAULT_VARIANT, CANDIDATE_VARIANT, TRACKING_CONTEXT_SCHEMA } from './constants';

export function getExperimentData(experimentName) {
  return get(window, ['gon', 'experiment', experimentName]);
}

export function getExperimentContexts(...experimentNames) {
  return experimentNames
    .map((name) => {
      const data = getExperimentData(name);
      return data && { schema: TRACKING_CONTEXT_SCHEMA, data };
    })
    .filter((context) => context);
}

export function isExperimentVariant(experimentName, variantName) {
  return getExperimentData(experimentName)?.variant === variantName;
}

export function getExperimentVariant(experimentName) {
  return getExperimentData(experimentName)?.variant || DEFAULT_VARIANT;
}

export function experiment(experimentName, variants) {
  const variant = getExperimentVariant(experimentName);

  switch (variant) {
    case DEFAULT_VARIANT:
      return variants.use.call();
    case CANDIDATE_VARIANT:
      return variants.try.call();
    default:
      return variants[variant].call();
  }
}
