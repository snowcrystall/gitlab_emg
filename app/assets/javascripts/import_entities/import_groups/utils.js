import { NEW_NAME_FIELD } from './constants';

export function isNameValid(group, validationRegex) {
  return validationRegex.test(group.import_target[NEW_NAME_FIELD]);
}

export function getInvalidNameValidationMessage(group) {
  return group.validation_errors.find(({ field }) => field === NEW_NAME_FIELD)?.message;
}

export function isInvalid(group, validationRegex) {
  return Boolean(!isNameValid(group, validationRegex) || getInvalidNameValidationMessage(group));
}
