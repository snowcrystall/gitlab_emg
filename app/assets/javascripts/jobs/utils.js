import * as Sentry from '@sentry/browser';

/**
 * capture anything starting with http:// or https://
 *   https?:\/\/
 *
 * up until a disallowed character or whitespace
 *   [^"<>()\\^`{|}\s]+
 *
 * and a disallowed character or whitespace, including non-ending chars .,:;!?
 *   [^"<>()\\^`{|}\s.,:;!?]
 */
export const linkRegex = /(https?:\/\/[^"<>()\\^`{|}\s]+[^"<>()\\^`{|}\s.,:;!?])/g;
export default { linkRegex };

export const reportToSentry = (component, failureType) => {
  Sentry.withScope((scope) => {
    scope.setTag('component', component);
    Sentry.captureException(failureType);
  });
};
