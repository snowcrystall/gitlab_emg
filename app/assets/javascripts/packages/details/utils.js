import { TrackingActions } from './constants';

export const trackInstallationTabChange = {
  methods: {
    trackInstallationTabChange(tabIndex) {
      const action = tabIndex === 0 ? TrackingActions.INSTALLATION : TrackingActions.REGISTRY_SETUP;
      this.track(action, { label: this.trackingLabel });
    },
  },
};
