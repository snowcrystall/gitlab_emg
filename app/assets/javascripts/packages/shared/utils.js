import { s__ } from '~/locale';
import { PackageType, TrackingCategories } from './constants';

export const packageTypeToTrackCategory = (type) =>
  // eslint-disable-next-line @gitlab/require-i18n-strings
  `UI::${TrackingCategories[type]}`;

export const beautifyPath = (path) => (path ? path.split('/').join(' / ') : '');

export const getPackageTypeLabel = (packageType) => {
  switch (packageType) {
    case PackageType.CONAN:
      return s__('PackageRegistry|Conan');
    case PackageType.MAVEN:
      return s__('PackageRegistry|Maven');
    case PackageType.NPM:
      return s__('PackageRegistry|npm');
    case PackageType.NUGET:
      return s__('PackageRegistry|NuGet');
    case PackageType.PYPI:
      return s__('PackageRegistry|PyPI');
    case PackageType.RUBYGEMS:
      return s__('PackageRegistry|RubyGems');
    case PackageType.COMPOSER:
      return s__('PackageRegistry|Composer');
    case PackageType.GENERIC:
      return s__('PackageRegistry|Generic');
    case PackageType.DEBIAN:
      return s__('PackageRegistry|Debian');
    case PackageType.HELM:
      return s__('PackageRegistry|Helm');
    default:
      return null;
  }
};

export const getCommitLink = ({ project_path: projectPath, pipeline = {} }, isGroup = false) => {
  if (isGroup) {
    return `/${projectPath}/commit/${pipeline.sha}`;
  }

  return `../commit/${pipeline.sha}`;
};
