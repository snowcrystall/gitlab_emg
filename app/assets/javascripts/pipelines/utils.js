import * as Sentry from '@sentry/browser';
import { pickBy } from 'lodash';
import { createNodeDict } from './components/parsing_utils';
import { SUPPORTED_FILTER_PARAMETERS } from './constants';

export const validateParams = (params) => {
  return pickBy(params, (val, key) => SUPPORTED_FILTER_PARAMETERS.includes(key) && val);
};

/**
 * This function takes the stages array and transform it
 * into a hash where each key is a job name and the job data
 * is associated to that key.
 * @param {Array} stages
 * @returns {Object} - Hash of jobs
 */
export const createJobsHash = (stages = []) => {
  const nodes = stages.flatMap(({ groups }) => groups);
  return createNodeDict(nodes);
};

/**
 * This function takes the jobs hash generated by
 * `createJobsHash` function and returns an easier
 * structure to work with for needs relationship
 * where the key is the job name and the value is an
 * array of all the needs this job has recursively
 * (includes the needs of the needs)
 * @param {Object} jobs
 * @returns {Object} - Hash of jobs and array of needs
 */
export const generateJobNeedsDict = (jobs = {}) => {
  const arrOfJobNames = Object.keys(jobs);

  return arrOfJobNames.reduce((acc, value) => {
    const recursiveNeeds = (jobName) => {
      if (!jobs[jobName]?.needs) {
        return [];
      }

      return jobs[jobName].needs
        .reduce((needsAcc, job) => {
          // It's possible that a needs refer to an optional job
          // that is not defined in which case we don't add that entry
          if (!jobs[job]) {
            return needsAcc;
          }

          // If we already have the needs of a job in the accumulator,
          // then we use the memoized data instead of the recursive call
          // to save some performance.
          const newNeeds = acc[job] ?? recursiveNeeds(job);

          // In case it's a parallel job (size > 1), the name of the group
          // and the job will be different. This mean we also need to add the group name
          // to the list of `needs` to ensure we can properly reference it.
          const group = jobs[job];
          if (group.size > 1) {
            return [...needsAcc, job, group.name, newNeeds];
          }

          return [...needsAcc, job, newNeeds];
        }, [])
        .flat(Infinity);
    };

    // To ensure we don't have duplicates job relationship when 2 jobs
    // needed by another both depends on the same jobs, we remove any
    // duplicates from the array.
    const uniqueValues = Array.from(new Set(recursiveNeeds(value)));

    return { ...acc, [value]: uniqueValues };
  }, {});
};

export const reportToSentry = (component, failureType) => {
  Sentry.withScope((scope) => {
    scope.setTag('component', component);
    Sentry.captureException(failureType);
  });
};

export const reportMessageToSentry = (component, message, context) => {
  Sentry.withScope((scope) => {
    // eslint-disable-next-line @gitlab/require-i18n-strings
    scope.setContext('Vue data', context);
    scope.setTag('component', component);
    Sentry.captureMessage(message);
  });
};
