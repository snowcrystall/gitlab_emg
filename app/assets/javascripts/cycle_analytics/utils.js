import dateFormat from 'dateformat';
import { unescape } from 'lodash';
import { dateFormats } from '~/analytics/shared/constants';
import { hideFlash } from '~/flash';
import { sanitize } from '~/lib/dompurify';
import { roundToNearestHalf } from '~/lib/utils/common_utils';
import { getDateInPast } from '~/lib/utils/datetime/date_calculation_utility';
import { parseSeconds } from '~/lib/utils/datetime_utility';
import { slugify } from '~/lib/utils/text_utility';
import { s__, sprintf } from '../locale';

export const removeFlash = (type = 'alert') => {
  const flashEl = document.querySelector(`.flash-${type}`);
  if (flashEl) {
    hideFlash(flashEl);
  }
};

/**
 * Takes the stages and median data, combined with the selected stage, to build an
 * array which is formatted to proivde the data required for the path navigation.
 *
 * @param {Array} stages - The stages available to the group / project
 * @param {Object} medians - The median values for the stages available to the group / project
 * @param {Object} stageCounts - The total item count for the stages available
 * @param {Object} selectedStage - The currently selected stage
 * @returns {Array} An array of stages formatted with data required for the path navigation
 */
export const transformStagesForPathNavigation = ({
  stages,
  medians,
  stageCounts = {},
  selectedStage,
}) => {
  const formattedStages = stages.map((stage) => {
    return {
      metric: medians[stage?.id],
      selected: stage?.id === selectedStage?.id, // Also could null === null cause an issue here?
      stageCount: stageCounts && stageCounts[stage?.id],
      icon: null,
      ...stage,
    };
  });

  return formattedStages;
};

export const timeSummaryForPathNavigation = ({ seconds, hours, days, minutes, weeks, months }) => {
  if (months) {
    return sprintf(s__('ValueStreamAnalytics|%{value}M'), {
      value: roundToNearestHalf(months),
    });
  } else if (weeks) {
    return sprintf(s__('ValueStreamAnalytics|%{value}w'), {
      value: roundToNearestHalf(weeks),
    });
  } else if (days) {
    return sprintf(s__('ValueStreamAnalytics|%{value}d'), {
      value: roundToNearestHalf(days),
    });
  } else if (hours) {
    return sprintf(s__('ValueStreamAnalytics|%{value}h'), { value: hours });
  } else if (minutes) {
    return sprintf(s__('ValueStreamAnalytics|%{value}m'), { value: minutes });
  } else if (seconds) {
    return unescape(sanitize(s__('ValueStreamAnalytics|&lt;1m'), { ALLOWED_TAGS: [] }));
  }
  return '-';
};

/**
 * Takes a raw median value in seconds and converts it to a string representation
 * ie. converts 172800 => 2d (2 days)
 *
 * @param {Number} Median - The number of seconds for the median calculation
 * @returns {String} String representation ie 2w
 */
export const medianTimeToParsedSeconds = (value) =>
  timeSummaryForPathNavigation({
    ...parseSeconds(value, { daysPerWeek: 7, hoursPerDay: 24 }),
    seconds: value,
  });

/**
 * Takes the raw median value arrays and converts them into a useful object
 * containing the string for display in the path navigation
 * ie. converts [{ id: 'test', value: 172800 }] => { 'test': '2d' }
 *
 * @param {Array} Medians - Array of stage median objects, each contains a `id`, `value` and `error`
 * @returns {Object} Returns key value pair with the stage name and its display median value
 */
export const formatMedianValues = (medians = []) =>
  medians.reduce((acc, { id, value = 0 }) => {
    return {
      ...acc,
      [id]: value ? medianTimeToParsedSeconds(value) : '-',
    };
  }, {});

export const filterStagesByHiddenStatus = (stages = [], isHidden = true) =>
  stages.filter(({ hidden = false }) => hidden === isHidden);

const toIsoFormat = (d) => dateFormat(d, dateFormats.isoDate);

/**
 * Takes an integer specifying the number of days to subtract
 * from the date specified will return the 2 dates, formatted as ISO dates
 *
 * @param {Number} daysInPast - Number of days in the past to subtract
 * @param {Date} [today=new Date] - Date to subtract days from, defaults to today
 * @returns {Object} Returns 'now' and the 'past' date formatted as ISO dates
 */
export const calculateFormattedDayInPast = (daysInPast, today = new Date()) => {
  return {
    now: toIsoFormat(today),
    past: toIsoFormat(getDateInPast(today, daysInPast)),
  };
};

/**
 * @typedef {Object} MetricData
 * @property {String} title - Title of the metric measured
 * @property {String} value - String representing the decimal point value, e.g '1.5'
 * @property {String} [unit] - String representing the decimal point value, e.g '1.5'
 *
 * @typedef {Object} TransformedMetricData
 * @property {String} label - Title of the metric measured
 * @property {String} value - String representing the decimal point value, e.g '1.5'
 * @property {String} key - Slugified string based on the 'title'
 * @property {String} description - String to display for a description
 * @property {String} unit - String representing the decimal point value, e.g '1.5'
 */

/**
 * Prepares metric data to be rendered in the metric_card component
 *
 * @param {MetricData[]} data - The metric data to be rendered
 * @param {Object} popoverContent - Key value pair of data to display in the popover
 * @returns {TransformedMetricData[]} An array of metrics ready to render in the metric_card
 */

export const prepareTimeMetricsData = (data = [], popoverContent = {}) =>
  data.map(({ title: label, ...rest }) => {
    const key = slugify(label);
    return {
      ...rest,
      label,
      key,
      description: popoverContent[key]?.description || '',
    };
  });
