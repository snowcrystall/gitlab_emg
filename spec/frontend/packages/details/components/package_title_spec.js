import { GlBreakpointInstance } from '@gitlab/ui/dist/utils';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import PackageTitle from '~/packages/details/components/package_title.vue';
import PackageTags from '~/packages/shared/components/package_tags.vue';
import TitleArea from '~/vue_shared/components/registry/title_area.vue';
import {
  conanPackage,
  mavenFiles,
  mavenPackage,
  mockTags,
  npmFiles,
  npmPackage,
  nugetPackage,
} from '../../mock_data';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('PackageTitle', () => {
  let wrapper;
  let store;

  function createComponent({
    packageEntity = mavenPackage,
    packageFiles = mavenFiles,
    icon = null,
  } = {}) {
    store = new Vuex.Store({
      state: {
        packageEntity,
        packageFiles,
      },
      getters: {
        packageTypeDisplay: ({ packageEntity: { package_type: type } }) => type,
        packagePipeline: ({ packageEntity: { pipeline = null } }) => pipeline,
        packageIcon: () => icon,
      },
    });

    wrapper = shallowMount(PackageTitle, {
      localVue,
      store,
      stubs: {
        TitleArea,
      },
    });
    return wrapper.vm.$nextTick();
  }

  const findTitleArea = () => wrapper.find(TitleArea);
  const packageType = () => wrapper.find('[data-testid="package-type"]');
  const packageSize = () => wrapper.find('[data-testid="package-size"]');
  const pipelineProject = () => wrapper.find('[data-testid="pipeline-project"]');
  const packageRef = () => wrapper.find('[data-testid="package-ref"]');
  const packageTags = () => wrapper.find(PackageTags);
  const packageBadges = () => wrapper.findAll('[data-testid="tag-badge"]');

  afterEach(() => {
    wrapper.destroy();
  });

  describe('renders', () => {
    it('without tags', async () => {
      await createComponent();

      expect(wrapper.element).toMatchSnapshot();
    });

    it('with tags', async () => {
      await createComponent({ packageEntity: { ...mavenPackage, tags: mockTags } });

      expect(wrapper.element).toMatchSnapshot();
    });

    it('with tags on mobile', async () => {
      jest.spyOn(GlBreakpointInstance, 'isDesktop').mockReturnValue(false);
      await createComponent({ packageEntity: { ...mavenPackage, tags: mockTags } });
      await wrapper.vm.$nextTick();

      expect(packageBadges()).toHaveLength(mockTags.length);
    });
  });

  describe('package title', () => {
    it('is correctly bound', async () => {
      await createComponent();

      expect(findTitleArea().props('title')).toBe('Test package');
    });
  });

  describe('package icon', () => {
    const fakeSrc = 'a-fake-src';

    it('binds an icon when provided one from vuex', async () => {
      await createComponent({ icon: fakeSrc });

      expect(findTitleArea().props('avatar')).toBe(fakeSrc);
    });

    it('do not binds an icon when not provided one', async () => {
      await createComponent();

      expect(findTitleArea().props('avatar')).toBe(null);
    });
  });

  describe.each`
    packageEntity   | text
    ${conanPackage} | ${'conan'}
    ${mavenPackage} | ${'maven'}
    ${npmPackage}   | ${'npm'}
    ${nugetPackage} | ${'nuget'}
  `(`package type`, ({ packageEntity, text }) => {
    beforeEach(() => createComponent({ packageEntity }));

    it(`${packageEntity.package_type} should render from Vuex getters ${text}`, () => {
      expect(packageType().props()).toEqual(expect.objectContaining({ text, icon: 'package' }));
    });
  });

  describe('calculates the package size', () => {
    it('correctly calculates when there is only 1 file', async () => {
      await createComponent({ packageEntity: npmPackage, packageFiles: npmFiles });

      expect(packageSize().props()).toMatchObject({ text: '200 bytes', icon: 'disk' });
    });

    it('correctly calulates when there are multiple files', async () => {
      await createComponent();

      expect(packageSize().props('text')).toBe('300 bytes');
    });
  });

  describe('package tags', () => {
    it('displays the package-tags component when the package has tags', async () => {
      await createComponent({
        packageEntity: {
          ...npmPackage,
          tags: mockTags,
        },
      });

      expect(packageTags().exists()).toBe(true);
    });

    it('does not display the package-tags component when there are no tags', async () => {
      await createComponent();

      expect(packageTags().exists()).toBe(false);
    });
  });

  describe('package ref', () => {
    it('does not display the ref if missing', async () => {
      await createComponent();

      expect(packageRef().exists()).toBe(false);
    });

    it('correctly shows the package ref if there is one', async () => {
      await createComponent({ packageEntity: npmPackage });
      expect(packageRef().props()).toMatchObject({
        text: npmPackage.pipeline.ref,
        icon: 'branch',
      });
    });
  });

  describe('pipeline project', () => {
    it('does not display the project if missing', async () => {
      await createComponent();

      expect(pipelineProject().exists()).toBe(false);
    });

    it('correctly shows the pipeline project if there is one', async () => {
      await createComponent({ packageEntity: npmPackage });

      expect(pipelineProject().props()).toMatchObject({
        text: npmPackage.pipeline.project.name,
        icon: 'review-list',
        link: npmPackage.pipeline.project.web_url,
      });
    });
  });
});
