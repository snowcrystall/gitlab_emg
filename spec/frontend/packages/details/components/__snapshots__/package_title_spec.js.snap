// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PackageTitle renders with tags 1`] = `
<div
  class="gl-display-flex gl-flex-direction-column"
  data-qa-selector="package_title"
>
  <div
    class="gl-display-flex gl-justify-content-space-between gl-py-3"
  >
    <div
      class="gl-flex-direction-column gl-flex-grow-1"
    >
      <div
        class="gl-display-flex"
      >
        <!---->
         
        <div
          class="gl-display-flex gl-flex-direction-column"
        >
          <h1
            class="gl-font-size-h1 gl-mt-3 gl-mb-2"
            data-testid="title"
          >
            Test package
          </h1>
           
          <div
            class="gl-display-flex gl-align-items-center gl-text-gray-500 gl-mt-1"
          >
            <gl-icon-stub
              class="gl-mr-3"
              name="eye"
              size="16"
            />
             
            <gl-sprintf-stub
              message="v%{version} published %{timeAgo}"
            />
          </div>
        </div>
      </div>
       
      <div
        class="gl-display-flex gl-flex-wrap gl-align-items-center gl-mt-3"
      >
        <div
          class="gl-display-flex gl-align-items-center gl-mr-5"
        >
          <metadata-item-stub
            data-testid="package-type"
            icon="package"
            link=""
            size="s"
            text="maven"
            texttooltip=""
          />
        </div>
        <div
          class="gl-display-flex gl-align-items-center gl-mr-5"
        >
          <metadata-item-stub
            data-testid="package-size"
            icon="disk"
            link=""
            size="s"
            text="300 bytes"
            texttooltip=""
          />
        </div>
        <div
          class="gl-display-flex gl-align-items-center gl-mr-5"
        >
          <package-tags-stub
            hidelabel="true"
            tagdisplaylimit="2"
            tags="[object Object],[object Object],[object Object],[object Object]"
          />
        </div>
      </div>
    </div>
     
    <!---->
  </div>
   
  <p />
</div>
`;

exports[`PackageTitle renders without tags 1`] = `
<div
  class="gl-display-flex gl-flex-direction-column"
  data-qa-selector="package_title"
>
  <div
    class="gl-display-flex gl-justify-content-space-between gl-py-3"
  >
    <div
      class="gl-flex-direction-column gl-flex-grow-1"
    >
      <div
        class="gl-display-flex"
      >
        <!---->
         
        <div
          class="gl-display-flex gl-flex-direction-column"
        >
          <h1
            class="gl-font-size-h1 gl-mt-3 gl-mb-2"
            data-testid="title"
          >
            Test package
          </h1>
           
          <div
            class="gl-display-flex gl-align-items-center gl-text-gray-500 gl-mt-1"
          >
            <gl-icon-stub
              class="gl-mr-3"
              name="eye"
              size="16"
            />
             
            <gl-sprintf-stub
              message="v%{version} published %{timeAgo}"
            />
          </div>
        </div>
      </div>
       
      <div
        class="gl-display-flex gl-flex-wrap gl-align-items-center gl-mt-3"
      >
        <div
          class="gl-display-flex gl-align-items-center gl-mr-5"
        >
          <metadata-item-stub
            data-testid="package-type"
            icon="package"
            link=""
            size="s"
            text="maven"
            texttooltip=""
          />
        </div>
        <div
          class="gl-display-flex gl-align-items-center gl-mr-5"
        >
          <metadata-item-stub
            data-testid="package-size"
            icon="disk"
            link=""
            size="s"
            text="300 bytes"
            texttooltip=""
          />
        </div>
      </div>
    </div>
     
    <!---->
  </div>
   
  <p />
</div>
`;
